### PR TITLE
Provide validation for data model for #switch section

### LINKED ISSUE #689: Validation for data model for #switch section
See https://quarkus.io/guides/qute-reference#when_section

Given:

```html
{#switch person.name}
  {#case 'John'} 
    Hey John!
  {#case 'Mary'}
    Hey Mary!
  {#case 123}
{/switch}
```

Here person.name is a String, but last #case (which i sequals to 123) is a number, we should higlight 123 as an error to say that it is an unexpeced type.

We should support too enum like

```html
{#when machine.status}
  {#is ON}
    It's running. 
  {#is in OFF BROKEN}
    It's broken or OFF. 
  {#is in BAD_ENUM_VAUE}
{/when}
```

BAD_ENUM_VAUE must be higlighted as en error. Please note that #is can have several values (OFF BROKEN)

### PR DESCRIPTION
**Pull request by JessicaJHee**
Fixes #689 

Signed-off-by: Jessica He <jhe@redhat.com>

### CODE REVIEW

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java [hunk 1]
```diff
@@ -1090,4 +1100,52 @@ private ResolvedJavaTypeInfo validateIterable(Part part, Section ownerSection,
 		}
 		return resolvedJavaType;
 	}
-}
+
+	private void validateSwitchWhenPart(JavaMethodInfo method, JavaMemberInfo member, Section ownerSection,
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java:1104]: JavaMethodInfo is a JavaMemeberInfo so I think you can remove method parzmeter znd use only member parameter

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java [hunk 2]
```diff
@@ -1090,4 +1100,52 @@ private ResolvedJavaTypeInfo validateIterable(Part part, Section ownerSection,
...
+					value = ((IsSection) node).getParameters();
+				}
+				for (Parameter child : value) {
+					String javaTypeChild = LiteralSupport.getLiteralJavaType(child.getValue());
+					// Case of Enum, check the Enum value is valid
+					if (javaTypeChild == null) {
+						String childName = child.getName();
+						ResolvedJavaTypeInfo javaTypeChildInfo = resolveJavaType(javaTypeParent, projectUri,
+								resolvingJavaTypeContext);
+						if (!"in".equals(childName)
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java:1134]: It seems that you don't take care of this usecase:

 * `{#is in}` : here `in` is the Java Type (not the in keyword).
`{#is in in on}`  : here the first `in` is the keyword and teh second  is the Java Type (not the in keyword).

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java [hunk 3]
```diff
@@ -1090,4 +1100,52 @@ private ResolvedJavaTypeInfo validateIterable(Part part, Section ownerSection,
...
+				if (ownerSection.getSectionKind() == SectionKind.SWITCH) {
+					value = ((CaseSection) node).getParameters();
+				} else if (ownerSection.getSectionKind() == SectionKind.WHEN) {
+					value = ((IsSection) node).getParameters();
+				}
+				for (Parameter child : value) {
+					String javaTypeChild = LiteralSupport.getLiteralJavaType(child.getValue());
+					// Case of Enum, check the Enum value is valid
+					if (javaTypeChild == null) {
+						String childName = child.getName();
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java:1131]: Could you add a sample in comment.

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java (hunk 2 above)
Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java:1134]: Could you add a sample in comment.

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java [hunk 4]
```diff
@@ -1090,4 +1100,52 @@ private ResolvedJavaTypeInfo validateIterable(Part part, Section ownerSection,
...
+								resolvingJavaTypeContext);
+						if (!"in".equals(childName)
+								&& javaCache.findMember(javaTypeChildInfo, childName, projectUri) == null) {
+							Range range = QutePositionUtility.createRange(child);
+							Diagnostic diagnostic = createDiagnostic(range, DiagnosticSeverity.Error,
+									QuteErrorCode.UnexpectedEnumValue, childName, partName, javaTypeParent);
+							diagnostics.add(diagnostic);
+						}
+					} else if (!javaTypeParent.equals(javaTypeChild)) {
+						Range range = QutePositionUtility.createRange(child);
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java:1142]: Could you add a sample in comment.

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java [hunk 5]
```diff
@@ -1090,4 +1100,64 @@ private ResolvedJavaTypeInfo validateIterable(Part part, Section ownerSection,
 		}
 		return resolvedJavaType;
 	}
-}
+
+	private void validateSwitchWhenPart(JavaMemberInfo member, Section ownerSection, String projectUri,
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java:1104]: Remove methodPart parameter and replace PropertyPart parameter with Part part.

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java [hunk 6]
```diff
@@ -1090,4 +1100,64 @@ private ResolvedJavaTypeInfo validateIterable(Part part, Section ownerSection,
...
+			partName = methodPart.toString();
+		} else if (propertyPart != null) {
+			partName = propertyPart.toString();
+		}
+		List<Node> children = ownerSection.getChildren();
+		// Iterate through children and ensure the type of the #case or #is parts are
+		// the same as the parent section
+		for (Node node : children) {
+			if (node.getKind() == NodeKind.Section) {
+				List<Parameter> value = new ArrayList<Parameter>();
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java:1119]: Here I will define

Section childSection = (Section) node;

And I will check with childSection.getSectionKind if it is a  #is or#case

And you could call

value = childSection.getParameters()

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java [hunk 7]
```diff
@@ -1090,4 +1100,64 @@ private ResolvedJavaTypeInfo validateIterable(Part part, Section ownerSection,
...
+				if (ownerSection.getSectionKind() == SectionKind.SWITCH) {
+					value = ((CaseSection) node).getParameters();
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java:1121]: Here you can have ClassCastException if you your section is not a #case (ex : #is). The main idea is to check if the current child node is a #is or #case otherwise you do nothing. I suggest that you do:

I suggest that you do:

```java
for (Node childNode : children) {
	if (childNode.getKind() == NodeKind.Section) {
		Section childSection = (Section) childNode;
		if (!(childSection.getSectionKind() == SectionKind.CASE || childSection.getSectionKind() == SectionKind.IS)) {
			// the current child section is not a #id or #case, ignore it 
			continue;
		}
		List<Parameter> value = childSection.getParameters();
```

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java [hunk 8]
```diff
@@ -1090,4 +1100,64 @@ private ResolvedJavaTypeInfo validateIterable(Part part, Section ownerSection,
...
+		String partName = "";
+		if (methodPart != null) {
+			partName = methodPart.toString();
+		} else if (propertyPart != null) {
+			partName = propertyPart.toString();
+		}
+		List<Node> children = ownerSection.getChildren();
+		// Iterate through children and ensure the type of the #case or #is parts are
+		// the same as the parent section
+		for (Node node : children) {
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java:1117]: rename node with `childNode`

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java (hunk 6 above)
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java:1119]: rename value with `parameters`

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java [hunk 9]
```diff
@@ -1090,4 +1100,64 @@ private ResolvedJavaTypeInfo validateIterable(Part part, Section ownerSection,
...
+			if (node.getKind() == NodeKind.Section) {
+				List<Parameter> value = new ArrayList<Parameter>();
+				if (ownerSection.getSectionKind() == SectionKind.SWITCH) {
+					value = ((CaseSection) node).getParameters();
+				} else if (ownerSection.getSectionKind() == SectionKind.WHEN) {
+					value = ((IsSection) node).getParameters();
+				}
+				boolean isKeywordIn = true;
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java:1125]: You can simplify the in keyword test by checking if the first parameter is "in" and if there are parameters size > 1.

Here the code I suggest you:

```java
List<Parameter> parameters = childSection.getParameters();
for (int i = 0; i < parameters.size(); i++) {
	Parameter parameter = parameters.get(i);
	String parameterName = parameter.getName();
	if (i == 0 && "in".equals(parameterName) && parameters.size() > 1) {
		// In the of 'in', the first is allowed as a keyword, second must be checked
		// Ex: {#is in in BROKEN}
		continue;
	}
	String javaTypeChild = LiteralSupport.getLiteralJavaType(parameter.getValue());
	if (javaTypeChild == null) {
		// Case of Enum since it is not a literal java type
		// {#when machine.status} --> machine.status is has values ON and OFF
		// {#is ON} --> valid enum value
		// {#is in BAD_ENUM_VAUE} --> Error: unexpected enum value
		// {/when}
		ResolvedJavaTypeInfo javaTypeChildInfo = resolveJavaType(javaTypeParent, projectUri,
				resolvingJavaTypeContext);
		if (javaCache.findMember(javaTypeChildInfo, parameterName, projectUri) == null) {
			Range range = QutePositionUtility.createRange(parameter);
			Diagnostic diagnostic = createDiagnostic(range, DiagnosticSeverity.Error,
					QuteErrorCode.UnexpectedEnumValue, parameterName, partName, javaTypeParent);
			diagnostics.add(diagnostic);
		}
	} else if (!javaTypeParent.equals(javaTypeChild)) {
		// Case where the java type of expression of child is different from parent
		// Ex: {#switch person.name} --> person.name is of type String
		// {#case 'John'} --> valid expression type of String
		// Hey John!
		// {#case 123} --> Error: unexpected type, expected String but was Integer
		// {/switch}
		Range range = QutePositionUtility.createRange(parameter);
		Diagnostic diagnostic = createDiagnostic(range, DiagnosticSeverity.Error,
				QuteErrorCode.UnexpectedMemberType, javaTypeChild, partName, javaTypeParent);
		diagnostics.add(diagnostic);
	}
}

```

File: qute.ls/com.redhat.qute.ls/src/test/java/com/redhat/qute/services/diagnostics/QuteDiagnosticsInExpressionTest.java [hunk 10]
```diff
@@ -706,6 +706,100 @@ public void parameterSuperType() {
 						DiagnosticSeverity.Error));
 	}
 
+	@Test
+	public void switchExpressionTypeMismatchSwitch() {
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/test/java/com/redhat/qute/services/diagnostics/QuteDiagnosticsInExpressionTest.java:710]: Move your tests in a new class `QuteDiagnosticsInExpressionWithSwitchWhenSectionTest`

**Comment by angelozerr** [comment]
You don't take  care of object part like:

`{#switch name}`

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java (hunk 5 above)
Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java:1104]: Im sorry @JessicaJHee I dont see the fix with just having part parameter?

File: qute.ls/com.redhat.qute.ls/src/test/java/com/redhat/qute/services/diagnostics/QuteDiagnosticsInExpressionWithSwitchWhenSectionTest.java [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+	}
+	
+	@Test
+	public void switchExpressionTypeMismatchSectionWhen() {
+		String template = "{@java.lang.String name}\r\n" + //
+				"		{#when name}\r\n" + //
+				"		{#case \"Apple\"}\r\n" + //
+				"		{/when}";
+		testDiagnosticsFor(template, //
+				d(1, 3, 1, 8, QuteErrorCode.MismatchingSection,
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/test/java/com/redhat/qute/services/diagnostics/QuteDiagnosticsInExpressionWithSwitchWhenSectionTest.java:72]: Are you sure that it is forbidden? I suggest that you try to write a qute template with this usecase and rendez the template and see there are some errors

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java [hunk 12]
```diff
@@ -265,6 +277,24 @@ private void validateDataModel(Node parent, Template template, QuteValidationSet
 				case INCLUDE:
 					validateIncludeSection((IncludeSection) section, diagnostics);
 					break;
+				case CASE:
+				case IS:
+					Section parentSection = section.getParentSection();
+					if (!isMatchSwitchWhenSection(parentSection, section)) {
+						Range range = QutePositionUtility.createRange(parentSection.getStartTagNameOpenOffset(),
+								parentSection.getStartTagNameCloseOffset(), template);
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java:285]: At first I suggest that you use QutePositionUtility.selectStartTag

After that I think it is better to set the error in #case or #if section start tag 
Instead of parent section. Indead if your template is just

{#case \"Apple\"}

You should have à NPE error because parentSection will be null.

Please use section instead of parentSection and pleqse write test with just 
{#case \"Apple\"}

File: qute.ls/com.redhat.qute.ls/src/test/java/com/redhat/qute/services/diagnostics/QuteDiagnosticsInExpressionWithSwitchWhenSectionTest.java [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+				"		{/switch}";
+		testDiagnosticsFor(template, //
+				d(3, 9, 3, 12, QuteErrorCode.UnexpectedMemberType,
+						"Unexpected type `java.lang.Integer` in `item.name`. Expected `java.lang.String`.",
+						DiagnosticSeverity.Error));
+	}
+	
+	@Test
+	public void switchExpressionTypeMismatchObjectPart() {
+		String template = "{@java.lang.String name}\r\n" + //
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/test/java/com/redhat/qute/services/diagnostics/QuteDiagnosticsInExpressionWithSwitchWhenSectionTest.java:42]: It should be nice to have the same test but with 
{#case XXXX}

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/parser/template/Section.java [hunk 14]
```diff
@@ -421,6 +421,11 @@ public Expression getExpressionParameter() {
 		return new ExpressionParameter(getStartParametersOffset(), getEndParametersOffset(), this);
 	}
 
+	public String getExpressionContent() {
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/parser/template/Section.java:424]: Add javadoc

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java [hunk 15]
```diff
@@ -368,7 +398,18 @@ private static void validateSectionTag(Section section, Template template,
...
+				|| sectionKind == SectionKind.SWITCH || sectionKind == SectionKind.WHEN;
+	}
+
+	private static boolean isMatchSwitchWhenSection(Section parentSection, Section childSection) {
+		if (parentSection == null || childSection == null) {
+			return false;
+		}
+		SectionKind childSectionKind = childSection.getSectionKind();
+		SectionKind parentSectionKind = parentSection.getSectionKind();
+		return ((parentSectionKind == SectionKind.SWITCH && childSectionKind == SectionKind.CASE)
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java:411]: Are sure that we should check that #case works only with #switch. I tthink that you  an use #case with #when
I suggest that yiu try that by rendering à qute template with this usecase

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java [hunk 16]
```diff
@@ -1090,4 +1131,47 @@ private ResolvedJavaTypeInfo validateIterable(Part part, Section ownerSection,
...
+			ResolvedJavaTypeInfo switchWhenType, String projectUri, ResolvingJavaTypeContext resolvingJavaTypeContext,
+			List<Diagnostic> diagnostics) {
+		String expression = whenSwitchSection.getExpressionContent();
+		List<Parameter> parameters = caseIsSection.getParameters();
+		// Iterate through children and ensure the type of the #case or #is parts are
+		// the same as the parent section
+		for (int i = 0; i < parameters.size(); i++) {
+			Parameter parameter = parameters.get(i);
+			String parameterName = parameter.getName();
+			if (i == 0 && "in".equals(parameterName) && parameters.size() > 1) {
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java:1145]: Replace with 

`if (i == 0 && parameters.size() > 1) {`

Indeed, as we can have several parameters (in, eq, etc, see https://github.com/quarkusio/quarkus/blob/9366bc89ad69ffd4ebef003b31b0e7cff46d8689/independent-projects/qute/core/src/main/java/io/quarkus/qute/WhenSectionHelper.java#L174)  

we need to take care of and it foolows the same logic than the real Qute parser:

https://github.com/quarkusio/quarkus/blob/9366bc89ad69ffd4ebef003b31b0e7cff46d8689/independent-projects/qute/core/src/main/java/io/quarkus/qute/WhenSectionHelper.java#L300

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/parser/template/Section.java (hunk 14 above)
Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/parser/template/Section.java:424]: I set to unresolved since there is no Javadoc

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java (hunk 15 above)
Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java:411]: After playing with Qute template #switch is the sam ething than #when and #is is the same thing than #case, so you can mix them. For instance this template is valid:

```
{#switch "A"}
	{#is "A"}
		A
	{#case "B"}
		B
{/switch}
```

The test chould be just: `sectionKind == SectionKind.SWITCH || sectionKind == SectionKind.WHEN` and you don't need childSection.

File: qute.ls/com.redhat.qute.ls/src/test/java/com/redhat/qute/services/diagnostics/QuteDiagnosticsInExpressionWithSwitchWhenSectionTest.java (hunk 11 above)
Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/test/java/com/redhat/qute/services/diagnostics/QuteDiagnosticsInExpressionWithSwitchWhenSectionTest.java:72]: This template is valid. You can mix #case and #is in #switch or #when. Please don't report error in this test.

File: qute.ls/com.redhat.qute.ls/src/test/java/com/redhat/qute/services/diagnostics/QuteDiagnosticsInExpressionWithSwitchWhenSectionTest.java (hunk 13 above)
Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/test/java/com/redhat/qute/services/diagnostics/QuteDiagnosticsInExpressionWithSwitchWhenSectionTest.java:42]: Sorry I don't see this test.

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java (hunk 12 above)
Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java:285]: I reopen the conversation since it has been opened and not resolved.

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java [hunk 17]
```diff
@@ -265,6 +275,23 @@ private void validateDataModel(Node parent, Template template, QuteValidationSet
 				case INCLUDE:
 					validateIncludeSection((IncludeSection) section, diagnostics);
 					break;
+				case CASE:
+				case IS:
+					Section parentSection = section.getParentSection();
+					if (!isSwitchWhenSection(parentSection)) {
+						Range range = QutePositionUtility.selectStartTagName(section);
+						String expectedSection = section.getSectionKind() == SectionKind.CASE ? "switch" : "when";
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java:283]: case can be hosted in a switch/when section. So I will change the error message to say that.

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java [hunk 18]
```diff
@@ -1090,4 +1126,46 @@ private ResolvedJavaTypeInfo validateIterable(Part part, Section ownerSection,
 		}
 		return resolvedJavaType;
 	}
-}
+
+	private void validateCaseIsPart(Section caseIsSection, Section whenSwitchSection,
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java:1130]: Add Javadoc

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java [hunk 19]
```diff
@@ -265,6 +275,23 @@ private void validateDataModel(Node parent, Template template, QuteValidationSet
 				case INCLUDE:
 					validateIncludeSection((IncludeSection) section, diagnostics);
 					break;
+				case CASE:
+				case IS:
+					Section parentSection = section.getParentSection();
+					if (!isSwitchWhenSection(parentSection)) {
+						Range range = QutePositionUtility.selectStartTagName(section);
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java:282]: Add a comment by adding a simple usecase

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/diagnostics/QuteErrorCode.java [hunk 20]
```diff
@@ -62,6 +62,11 @@ public enum QuteErrorCode implements IQuteErrorCode {
 	MethodIgnoredByTemplateData(
 			"The method `{0}` of `{1}` Java type cannot be used in native image mode because Java type is annotated with `@TemplateData(ignore = {2})`."), //
 
+	// Error code for expression part
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/diagnostics/QuteErrorCode.java:65]: // Error code for expression part in #case  / #is section

Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/diagnostics/QuteErrorCode.java:65]: I think we should have an error code which says that it is for #case (or #is). I will rename: 

 * UnexpectedMemberType -> UnexpectedMemberTypeInCaseSection
 *  UnexpectedPartValue-> UnexpectedValueInCaseSection
 *  MismatchingSection-> InvalidParentInCaseSection

**Comment by angelozerr** [comment]
@JessicaJHee I have tested your PR with a real project by creating a Java enum file and it doesn't work. The reason is that `ResolvedJavaTypeInfo#getFields()` return an empty array. It is because on JDT side, fields for enum are not collected.

To fix this problem:

 * replace the line https://github.com/redhat-developer/quarkus-ls/blob/37bb5f95f533d72fb09c817a85ac634c3684adf5/qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/QuteSupportForTemplate.java#L369 with 
 
`if (isValidField(field, type)) {`

 * replace isValidField at https://github.com/redhat-developer/quarkus-ls/blob/37bb5f95f533d72fb09c817a85ac634c3684adf5/qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/QuteSupportForTemplate.java#L465 with 

```java
private static boolean isValidField(IField field, IType type) throws JavaModelException {
	if (type.isEnum()) {
		return true;
	}
	return Flags.isPublic(field.getFlags());
}
```

In other words Java enum fields are private. Once you did that, please write a test with the existing [StatusesEnum](https://github.com/redhat-developer/quarkus-ls/blob/37bb5f95f533d72fb09c817a85ac634c3684adf5/qute.jdt/com.redhat.qute.jdt.test/projects/maven/qute-quickstart/src/main/java/org/acme/qute/StatusesEnum.java#L6) in https://github.com/redhat-developer/quarkus-ls/blob/master/qute.jdt/com.redhat.qute.jdt.test/src/main/java/com/redhat/qute/jdt/template/TemplateGetResolvedJavaTypeTest.java to check that fields of ResolvedJavaType for StatusesEnum has ON and OFF.

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java [hunk 21]
```diff
@@ -368,7 +397,16 @@ private static void validateSectionTag(Section section, Template template,
 	private static boolean canChangeContext(Section section) {
 		SectionKind sectionKind = section.getSectionKind();
 		return sectionKind == SectionKind.EACH || sectionKind == SectionKind.FOR || sectionKind == SectionKind.LET
-				|| sectionKind == SectionKind.SET || sectionKind == SectionKind.WITH;
+				|| sectionKind == SectionKind.SET || sectionKind == SectionKind.WITH
+				|| sectionKind == SectionKind.SWITCH || sectionKind == SectionKind.WHEN;
+	}
+
+	private static boolean isSwitchWhenSection(Section section) {
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java:404]: The name `isSwitchWhenSection` can be confused, I think we should consider that we use switch and case for all code. I will rename `isSwitchWhenSection `to `isSwitchSection`

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java [hunk 22]
```diff
@@ -1090,4 +1128,56 @@ private ResolvedJavaTypeInfo validateIterable(Part part, Section ownerSection,
...
+	 * Validate the part in a case or in section.
+	 *
+	 * @param caseIsSection     the case or is section.
+	 * @param whenSwitchSection the parent switch or when section.
+	 * @param switchWhenType    the type of the switch or when section that is
+	 *                          expected in the case or in section.
+	 * @param projectUri        the project Uri.
+	 * @param diagnostics       the diagnostic list.
+	 */
+	private void validateCaseIsPart(Section caseIsSection, Section whenSwitchSection,
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java:1142]: I will rename `caseIsSection` with `caseSection`

Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java:1142]: Rename with validateCaseSectionParameters

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/diagnostics/QuteErrorCode.java [hunk 23]
```diff
@@ -62,6 +62,11 @@ public enum QuteErrorCode implements IQuteErrorCode {
 	MethodIgnoredByTemplateData(
 			"The method `{0}` of `{1}` Java type cannot be used in native image mode because Java type is annotated with `@TemplateData(ignore = {2})`."), //
 
+	// Error code for expression part in #case / #is section
+	UnexpectedMemberTypeCaseInSection("Unexpected type `{0}` in `{1}`. Expected `{2}`."),
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/diagnostics/QuteErrorCode.java:66]: Should not be `UnexpectedMemberTypeInCaseSection` ?

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/diagnostics/QuteErrorCode.java [hunk 24]
```diff
@@ -62,6 +62,11 @@ public enum QuteErrorCode implements IQuteErrorCode {
 	MethodIgnoredByTemplateData(
 			"The method `{0}` of `{1}` Java type cannot be used in native image mode because Java type is annotated with `@TemplateData(ignore = {2})`."), //
 
+	// Error code for expression part in #case / #is section
+	UnexpectedMemberTypeCaseInSection("Unexpected type `{0}` in `{1}`. Expected `{2}`."),
+	UnexpectedValueCaseInSection("Unexpected value `{0}` in `{1}`. Expected value of type `{2}`."),
+	InvalidParentCaseInSection("`{0}` section must be hosted in a #switch or #when section."),
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/diagnostics/QuteErrorCode.java:68]: Should not be `InvalidParentInCaseSection` ?

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/diagnostics/QuteErrorCode.java [hunk 25]
```diff
@@ -62,6 +62,11 @@ public enum QuteErrorCode implements IQuteErrorCode {
 	MethodIgnoredByTemplateData(
 			"The method `{0}` of `{1}` Java type cannot be used in native image mode because Java type is annotated with `@TemplateData(ignore = {2})`."), //
 
+	// Error code for expression part in #case / #is section
+	UnexpectedMemberTypeCaseInSection("Unexpected type `{0}` in `{1}`. Expected `{2}`."),
+	UnexpectedValueCaseInSection("Unexpected value `{0}` in `{1}`. Expected value of type `{2}`."),
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/diagnostics/QuteErrorCode.java:67]: Should not be `UnexpectedValueInCaseSection` ?

**Comment by angelozerr** [comment]
It works like a charm:

CaseValidation

Great job @JessicaJHee !
